runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs

```diff
@@ -35,32 +36,95 @@ pub fn parse_rlp_tx_to_action(
-    let validation_outcome = validate_tx_relayer_data(&tx, target, context, *expected_nonce)?;
+    let target_kind = validate_tx_relayer_data(&tx, target, context, *expected_nonce)?;
 
     // If the transaction is valid then increment the nonce to prevent replay
     *expected_nonce = expected_nonce.saturating_add(1);
 
-    let to = tx.to.ok_or(Error::User(UserError::EvmDeployDisallowed))?.raw();
-    let action = if to != context.current_address
-        && extract_address(target).map(|a| a == to).unwrap_or(false)
-    {
-        // If target is another Ethereum implicit account then the action
-        // must be a transfer (because EOAs are not contracts on Ethereum).
-        Action::Transfer { receiver_id: target.to_string(), yocto_near: 0 }
-    } else {
-        parse_tx_data(target, &tx, context)?
+    // The way an honest relayer assigns `target` is as follows:
+    // 1. If the Ethereum transaction payload represents a Near action then use the receiver_id,
+    // 2. If the payload looks like a supported Ethereum emulation then use the address registrar:
+    // 2.a. if the tx.to address is registered then use the associated account id,
+    // 2.b. otherwise, tx.to == target
+    // 3. Otherwise, tx.to == target
+    // Given this algorithm, the only way to have `TargetKind::EthImplicit` is in the
+    // following cases:
+    // I)   The Ethereum transaction payload is not parseable as a known action,
+    // II)  The payload is parsable as a Near action and the receiver_id is an implicit account
+    // III) The payload is parsable as a supported Ethereum emulation but the to address is
+    //      not registered in the address registrar.
+    // Therefore, to determine if the relayer is honest we must always parse the payload and
+    // we only need to check the registrar if the payload is parseable as an Ethereum emulation.
+
+    let (action, transaction_kind) = match parse_tx_data(target, &tx, context) {
+        Ok((action, ParsableTransactionKind::NearNativeAction)) => {
+            (action, TransactionKind::NearNativeAction)
+        }
+        Ok((action, ParsableTransactionKind::SelfNearNativeAction)) => {
+            if let TargetKind::EthImplicit(_) = target_kind {
+                // The calldata was parseable as a Near native action where the target
+                // should be the current account, but the target is some other wallet contract.
+                // This is technically allowed under the Ethereum standard for base token transfers
+                // (where any calldata can be used when sending tokens to another EOA), so we
+                // assume such a transfer must have been the user's intent. No address check is
+                // required in this case because no Near account other than the current account
+                // can be the receiver of these actions.
+                (
+                    Action::Transfer { receiver_id: target.to_string(), yocto_near: 0 },
```

Comment:
Why create a `Transfer` of 0 `yocto_near`? Same for the `Transfer` of 0 below.

Reply:
The action may look confusing, but it is not actually a transfer of 0. There is a distinction between the Action that is defined by the Solidity ABI accepted by the contract and the true Near action. The former is what is being defined here and it includes a `yocto_near` field which represents the amount of yoctoNEAR that cannot be included in the Ethereum transaction value field due to the different number of decimals in ETH compared to NEAR (18 vs 24). The 0 here means 0 unrepresentable yoctoNEAR, i.e. the total transferred amount comes from the Ethereum transaction. This conversion to the total amount [happens at the bottom](https://github.com/aurora-is-near/nearcore/blob/eth-wallet-nep-141-deposit/runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs#L124) of the `parse_rlp_tx_to_action` function.